src/tcpstream.c

```diff
@@ -517,26 +520,32 @@ int ntttcp_server_epoll(struct ntttcp_stream_server *ss)
 			}
 			/* handle data from an EXISTING client */
 			else {
-				bzero(buffer, ss->recv_buf_size);
-				bytes_to_be_read = ss->is_sync_thread ? 1 : ss->recv_buf_size;
-
-				/* got error or connection closed by client */
-				if ((nbytes = n_read(current_fd, buffer, bytes_to_be_read)) <= 0) {
-					if (nbytes == 0) {
-						ASPRINTF(&log, "socket closed: %d", i);
-						PRINT_DBG_FREE(log);
+				for (int max_io = 0; max_io < MAX_IO_PER_POLL; max_io++) {
+					bytes_to_be_read = ss->is_sync_thread ? 1 : ss->recv_buf_size;
+
+					/* got error or connection closed by client */
+					errno = 0;
+					nbytes = n_recv(current_fd, buffer, bytes_to_be_read);
+					if (nbytes <= 0) {
+						if (errno != EAGAIN) {
```

Comment:
This (errno == EAGAIN) has been handled in n_send(), right?

Reply:
I don't understand this question: this is handling the result of n_recv() rather than n_send(), and the EAGAIN case needs to be handled differently than a fatal error: EAGAIN breaks from the loop, but a fatal error requires the socket be closed.

Reply:
In n_recv() and n_send(), it processes for EAGAIN. See the code of line #22 of src/tcpstream.c.

Reply:
```
int n_recv(int fd, char *buffer, size_t total)
{
	register ssize_t rtn;
	register size_t left = total;

	while (left > 0) {
		rtn = recv(fd, buffer, left, 0);
		if (rtn < 0) {
			if (errno == EINTR || errno == EAGAIN) {
				break;
			} else {
				printf("socket read error: %d\n", errno);
				return ERROR_NETWORK_READ;
			}
		}
		else if (rtn == 0)
			break;

		left -= rtn;
		buffer += rtn;
	}

	return total - left;
}
```